src/sentry/integrations/source_code_management/repository.py

```diff
@@ -0,0 +1,200 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Collection, Mapping, Sequence
+from typing import Any
+
+import sentry_sdk
+
+from sentry.auth.exceptions import IdentityNotValid
+from sentry.integrations.services.integration import integration_service
+from sentry.integrations.services.integration.model import RpcIntegration
+from sentry.integrations.services.repository import RpcRepository, repository_service
+from sentry.models.identity import Identity
+from sentry.models.integrations.integration import Integration
+from sentry.models.repository import Repository
+from sentry.shared_integrations.client.base import BaseApiResponseX
+from sentry.shared_integrations.exceptions import ApiError, IntegrationError
+
+
+class RepositoryIntegration(ABC):
+    repo_search = False
+    model: RpcIntegration | Integration
+
+    # @abstractmethod
+    @property
+    def codeowners_locations(self) -> Sequence[str] | None:
+        """
+        A list of possible locations for the CODEOWNERS file.
+        """
+        return None
+
+    @abstractmethod
+    def get_client(self) -> RepositoryClient:
+        """Returns the client for the integration. The client must be a subclass of RepositoryClient."""
+        raise NotImplementedError
```

Comment:
Is there a way we could maybe pass a client instance or factory to the class instead of having the subclass method to create one?

Reply:
i've actually been typing some classes in relation to this and found that some of them have different signatures (e.g. VSTS's `get_client` takes an optional `base_url` parameter), would a factory or client instance be better in that case?

Reply:
i'm adding some logging to make sure we don't need to add an overloaded function to handle VSTS's case https://github.com/getsentry/sentry/pull/75277